### ISSUE TITLE
AppGW using host 

### ISSUE THREAD

**Issue by reza8iucs** [Azure/aca-landing-zone-accelerator#65]
Hi,

I was wondering what's the rational to pick the host name from backend ACA in the application gateway backend setting? isn't that against Microsoft's recommendation?  Why not using the self-signed cert in the backend ACA app?

I spent a few hours to install this repo to see exactly how you set that up but the custom domain was not setup in ACA.

![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/110741554/23ace51b-11d9-4dcc-b32a-f7a451b98e8a)

**Comment by thotheod** [comment]
Hi @reza8iucs. It is possible that some applications might fail if the reverse application proxy overrides the http host name, as it is [described here](https://learn.microsoft.com/en-us/azure/architecture/best-practices/host-name-preservation). However there are also cases that you might intentionally  want to do that, i.e. to avoid complexity of deploying split-horizon DNS  etc. 

If you need to preserve the host name, you can do it, and you have two options for that:
- **option 1**:  Deploy everything as is (with _Override with new host name_ set to true), and then use appGW's [Rewrite HTTP headers feature](https://learn.microsoft.com/en-us/azure/application-gateway/rewrite-http-headers-url#modify-a-redirection-url) to  change the backend's response
- **option 2**: add a [custom domain-with your own certificate](https://learn.microsoft.com/en-us/azure/container-apps/custom-domains-certificates) in the Azure Container App. To do that you need:
    - a custom domain name. If you don't have any, you can [buy an App Service domain](https://learn.microsoft.com/en-us/azure/app-service/manage-custom-dns-buy-domain)
    - a certificate matching your custom domain. You can get a free certificate from [let's encrypt](https://letsencrypt.org/). You can use Azure Key Vault to create a certificate request, and finally store the certificate securely. More details on that you can find in this [blog post](https://trstringer.com/azure-key-vault-lets-encrypt/). 
    
  Once you have those:
    -  Add an A record on you custom Domain DNS zone pointing to the public IP of the application gateway
    - edit the AppGW httpListener to add your certificate matching your custom domain (in my case named mscaelab.com) 
    
<img width="517" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/f14de244-30a1-4525-9788-074d123d0f57">

    - Then change the https backend setting, set override with new host name to No
    
<img width="788" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/70c3e3b2-20be-4d03-b4c8-6caac86fd4e5">

    - Go to your ACA, go to Custom Domains and click Add a custom domain. There select the Bring your own certificate as shown below:   
    
![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/8f569e32-85ef-479f-9ab4-36417382f888)

    - Then you need to do a domain validation. **IMPORTANT: ** You should not add the A record for the domain validation (as instructed in the portal see figure below), because this is the INTERNAL IP of the ACA Ingress, but for the same custom domain we need to resolve the Pip of the appGW. 
    
<img width="393" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/7bd0ddb0-2366-42e3-be12-4d95c3afcfff">

  Once the domain is validated, and the certificate correct you get a green icon with "Secured" status in the custom domains list. 
>**IMPORTANT NOTE**: You should not change the app Gateway backendpool. The backend target should still point to the ACA default domain (i.e. ca-simple-hello.gentlebeach-ff5dee74.northeurope.azurecontainerapps.io) and not the custom domain. The Aca default domain can be resolved from the appGW because we have already (the LZA script does that) configured a private DNS zone. The ACA has a custom domain as hostname, but is not (and should NOT) be resolved from the external world.  

I have tested the configuration described in option 2 above and you can check that it works here:  https://mscaelab.com/ ( I will destroy the deployment in a few days though)

**Comment by reza8iucs** [comment]
thanks for going the extra miles to set this up. When you did set it up with your own custom cert, did you change the AppGW backend setting with Override with new host name set to **No**?

if yes, can you do the exact same setup you did above with the new managed cert and instead of specifying the root of your domain use a subdomain like (https://hello.mscaelab.com) with CNAME and TXT records? we have been banging our head against the wall for 2 days trying to set this up in ACA!

**Comment by reza8iucs** [comment]
Hi @thotheod, 

Thanks for going the extra miles to set this up. When you did set it up with your own custom cert, did you change the AppGW backend setting with Override with new host name set to No?

if yes, can you do the exact same setup you did above with the new managed cert and instead of specifying the root of your domain use a subdomain like (https://hello.mscaelab.com/) with CNAME and TXT records? we have been banging our head against the wall for 2 days trying to set this up in ACA!

Thank you!

**Comment by thotheod** [comment]
Hi @reza8iucs . Yes I have changed the appGW backend setting with "Override with new host name" set to No. Even if you use the subdomain (i.e. hello.mscaelab.com) should not be different.  

Bear in mind, that in order to validate your domain **you need just to add the txt record**. The CNAME or A record, must point to the public IP of the AppGW, not the private IP of the ACA. Ignore what is suggested in the portal UX when you try to Add a custom domain and certificate for ACA. 

I have done the same for subdomain (testkv.mscaelab.com) and it works. 

In ordetr to isolate and detect the problem, break down the process. Use the jumpbox VM to test that the custom domain works fine (connect through bastion, and add the internal IP of the ACA LB in the C:\Windows\System32\drivers\etc\hosts file. )
Check in the browser of the jumpbox if the custom domain works. If it works internally thenm something is wrong in the AppGW configuration

Do you get Green status here? 
<img width="440" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/6db0030d-6a90-431e-a211-e83e65f3d061">

Does the broswer inside jump-box works with custom domain works? (local/internal IP set in hosts file): 
<img width="477" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/406436e2-c631-4a38-8317-88f4812f5404">

**Comment by reza8iucs** [comment]
Thanks @thotheod. Much appreciate!

Two Questions:
1) In the backend pool , are you using FGDN of the ACA app or IP of the ACA environment? 
2) Also, did you try this with your new "ACA Managed Certificate" offering which is in preview or for the subdomain you are still using BYOD custom domain?  

P.S. I am really hoping you document and publish this setup in the Docs. A lot of people can benefit from this like the same one that currently exists for AppGW and APIM internal mode [here](https://learn.microsoft.com/en-us/azure/api-management/api-management-howto-integrate-internal-vnet-appgateway). Just thinking loud here.

**Comment by thotheod** [comment]
Hi @reza8iucs, yes you can use the internal ACA IP, but in general it's better to use FQDNs.  If you use FQDN you need to use the default ACA fqdn (not the custom domain) i.e. something like: ca-simple-hello.gentlebeach-ff5dee74.northeurope.azurecontainerapps.io

No you cannot use the "_Managed certificate_", because as you see in the [first bullet here](https://learn.microsoft.com/en-us/azure/container-apps/custom-domains-managed-certificates?pivots=azure-portal#free-certificate-requirements), your http ingress needs to be publicly available, which in the case of ACA LZA, is not.

**Comment by reza8iucs** [comment]
Hi @thotheod 

When isolating the problem , I can confirm that I am able to hit the ACA app internally from the jumpbox by modifying the host file as per your instruction.  So I think ACA custom domain is ok...mind you that i had to add the custom domain first and do the binding later as validation was not going through.

When hitting it through the app gateway,however, the only way I am able to get it to work is by applying the following in the backend setting otherwise AppGW gives me a 500 error because of an unhealthy probe. Did you use a custom health prob in your setup? I was hoping to get away with this whiteout having to set "Override with new host name" to true but it doesnt work for me.

![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/110741554/77b1cec0-fcc5-4ee2-974e-e6a7ad36761c)

**Comment by thotheod** [comment]
Hi @reza8iucs ,
Yes I Have a custom probe, and yes I have switched "Override with new host name" to No
check screenshots of some AppGW settings below, I hope that helps: 

backend pool (internal IP works as well instead of FQDN, I have tested both
![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/7e039119-4766-4f8e-baa5-93be2ae3f7fa)


Backend Settings (https)
![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/99c290ec-7173-4f69-96ea-888c286ea350)
<img width="430" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/e383e760-b215-4481-ad2b-ad6584a78ffb">


httpListener:
<img width="518" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/f94fe165-5276-44fc-97f9-abc99845c8ca">

Routing Rules (Listener)
<img width="596" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/90e09901-2a72-4507-86d8-ebd0ef7bf563">

Routing Rules (Backend Targets)
<img width="584" alt="image" src="https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/d0464fc3-358a-4d5b-adab-9c102b0e93bd">

Custom probe (named: webProbe)
![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/23176972/e1a3dea5-eec6-40c4-8dfc-a9b7d317cadc)

**Comment by reza8iucs** [comment]
Hi @thotheod 

Thanks for including all the screenshots. 

Modified the custom health prob Not to pick the host from the backend and specified the host in the custom health prob exactly like yours and solved the 500 error. However , now I am getting 502 bad gateway from AppGW unless I set the " "Override with new host name" to Yes and then specify "Pick host name from backend target" 

![image](https://github.com/Azure/aca-landing-zone-accelerator/assets/110741554/dbf24a93-7826-4d95-9179-55528a83815f)

we checked all of your screenshots and have exact same setup as yours but for us ""Override with new host name" to No doesnt work. so strange! 

Custom domain on the ACA app side works flawlessly as i am able to hit it from the jumbox with host file change. It must be a firewall or UDR thing that presents the appgw to communicate with ACA with the host header is not overridden. I've never seen an issue like this before.

**Comment by reza8iucs** [comment]
Hi @thotheod 

We managed to resolve the issue with the instructions you provided above. THANK YOU!

For anyone else who is getting 502 error in hitting their ACA app behind appGW. There are a number of root causes that you might be getting this issue which is documented [here](https://learn.microsoft.com/en-us/azure/application-gateway/application-gateway-troubleshooting-502).  Our issue was the certificate used in  the ACA custom domain. Our certificate was issued for the root domain contoso.com , but the ACA app was setup to use hello.constoso.com. if that's the case you either have to use a wildcard certificate *.contoso.com or a certificate that's issued for that subdomain in our case hello.contoso.com. We used the same certificate in the AppGW listener as per @thotheod but you dont have to.
